src/05-derivatives/02-common-data-types.md

```diff
@@ -8,7 +8,7 @@ Template:
 <pipeline_name>/
     sub-<participant_label>/
         anat|func|dwi/
-        <source_keywords>[_space-<space>][_desc-<label>]_<suffix>.<ext>
+        <source_keywords>[_space-<space>][_res-<index>][_den-<label>][_desc-<label>]_<suffix>.<ext>
```

Comment:
The indices thing feels weird, as opposed to a (potentially) descriptive label, given that the indices don't mean anything.

Reply:
As mentioned below, indices allow for the developer to impose some explicit order they want.

Reply:
Where is the index made explicit, i.e. where is the lookup-table that allows me to see what index value "1" signifies? Is the lookup table a TSV or a JSON?

Is the label in `den-label` free to choose by the researcher, or is it limited to a controlled vocabulary?

Reply:
I believe the index look-up is in the sidecar, [specifically here](https://github.com/bids-standard/bids-specification/pull/301/files#diff-ee997314dc6a77ebab0d882ed7945467R77).
On my reading, it seems that `den` is free text (rather than a controlled vocabulary). I'm not sure what the controlled vocabulary would be, here.

Reply:
thanks @emdupre, that clarifies.

I think it would be better if both `res` and `den` would allow strings rather than digits. This allows them to carry explicit meaning (like "1mm", "2mm"). Digits to indicate indices suffer from some ambiguities, i.e. should they start Python/C style 0-offset or MATLAB/Fortran style 1-offset? And is it allowed to use them in non-successive order like 1, 3, 5, 7?  The developer can (and probably should) always start by reading the JSON, and subsequently loop over the keys of the dictionary.

Reply:
The proposal of an index for `res` here follows that of `run` (https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#the-run-entity), i.e., start at 1, non-successive order disallowed.

To find the actual meaning of each resolution, the user MUST look into the file format, to avoid that you say `res-1x1x1mm` but then the headers say that the voxel sizes are 0.7x0.7x0.7.

Having labels like "1mm" seemed pretty limited to me, and quite misleading - for instance if your sampling is oblique (i.e., a NIfTI file with a shearing affine).

The JSON description field is free-from to preempt using the dictionary to shadow this information in the file format.

Reply:
@oesteban - unlike `run` which has some order to it, `res` would not have any specific order.   if the intent is to define a mapping to something to something more interpretable, then the use of the `index` is for disambiguation. in this scenario, wouldn't `desc` suffice for disambiguation?

if `index` is not a **controlled** vocabulary, people would have to look up, to do search. i.e. a glob pattern would not necessarily work.

if the intent is simply to say this is a regularly sampled n-dgrid, perhaps then `res-700mm` would be ok, with number of dimensions not being important for disambiguation. this makes the value an interpretable whole number without decimals and with units (we can follow the common units terms). this does pose the additional constraint that the number should match whatever is in the file, but that is something a future validator could check.

Reply:
> wouldn't `desc` suffice for disambiguation?

Only when it is not already used for disambiguation of some other semantic. The concept of resolution/density on digital data [is fundamental](https://github.com/bids-standard/bids-specification/pull/301#issuecomment-522264904) and justifies extracting it from `desc`. In my opinion, `desc` is more about the actual content of the file and `res` tells you about the structure of the data. However, in my opinion, the actual resolution (regardless number of dimensions, or units) is a detail that is better kept with the file format (and unnecessary to fully understand that we have two or more resolutions of the same reality).

> if `index` is not a controlled vocabulary, people would have to look up, to do search. i.e. a glob pattern would not necessarily work.

Okay, you mean that you could have two `res-1` pointing to different actual resolutions? Well, I don't think that is a problem, as long as you don't give `res` more powers than: _there is another file with the same content but sampled differently_. If you write a JSON metadata dictionary at the top level to describe the indexes then you are enforcing that all files sharing the index of `res` MUST have the same resolution and that all the subfolders have all the resolution indexes when `res` is present.

My conflict with `res-700mm` is that, like `desc` talks more about the content, not the structure of the data (and I see these labels very impractical too).

These explanations are probably not sufficiently described in the pull request, happy to iterate over it and make them more clear (or change my mind if I'm missing the point).

Reply:
> you could have two res-1 pointing to different actual resolutions?

indeed. yes, we would need a validator that enforces this, and figures out inheritance principles and such mundane things.

on the more general note, i think the reason you want to include `res` is for disambiguation (i.e. the use case of a file in two different resolutions of MNI template/fsaverage/etc.,.), but the disambiguating info is the content (the resolution of the data), which is likely already in the file itself. so whether `res` talks resolution via an `index`, or not, it is really talking about the data !

i think you already agreed with me :)
> concept of resolution/density on digital data is fundamental and justifies extracting it from desc

Reply:
After thinking a bit more about this (and similar ideas from @robertoostenveld above) - would you agree in relaxing the value of `res` to be a `<label>` (and hence, consistent with `den-`, which is nice), but including the recommendation to use one-based, consecutive indexes?

Reply:
i'm still missing what the 1-based consecutive index recommendation really helps with, given that labels can be arbitrary. and how would you differentiate across datasets that follow the recommendation and those that don't?

Reply:
> i'm still missing what the 1-based consecutive index recommendation really helps with, given that labels can be arbitrary.

1. Gives you an idea of how many different resolutions there are with a quick look. Other options don't.
2. Allows the developers to impose an order they might be interested in (e.g., higher resolution first).
3. Preempts inconsistency between the label and the data (e.g., having `res-333mm` for a NIfTI with 2x2x2 or that has an affine with shearing)

> how would you differentiate across datasets that follow the recommendation and those that don't?

The recommendation does not allow you to remove the metadata in the JSON. Other than that, why would you need to know whether the dataset sticks with the recommendation or not?

Reply:
Umm - I'm a little confused - I also believe that it should be a `res-<label>`  and not `res-<index>`.

I simple alphabetical sort of labels is probably all the ordering we would need?

I really hate the idea of res "2x2x2mm" being associated with index 3...that seems way too confusing.

Reply:
I modified the PR and now it proposes `res-<label>`. However, I left a recommendation indicating the use of indexes. The question is whether that should be removed or not. I could also expand the [recommendation box in the docs](https://1193-150465237-gh.circle-artifacts.com/0/home/circleci/project/site/05-derivatives/02-common-data-types.html#processed-coregistered-andor-resampled-volumes) to include the three reasons I gave above.